### PR TITLE
Add (some) dispatchables

### PR DESCRIPTION
**Pull request by bedeho**
## No need to do a very detailed review, because some things will change here, just keep it high level, 

1. **I will fix and merge https://github.com/Joystream/substrate-forum-module/pull/3 later today or tomorrow, changes will not be big based on last review. Then I we can rebase this branch if decide to merge it after this first review.**

2. So I reduced the ambitions for this PR, it just adds

- [x] set_forum_sudo
- [x] create_category
- ~update_category~
- ~create_thread~
- ~moderate_thread~
- ~add_post~
- ~edit_post_text~
- ~moderate_post~

with quite a few tests missing on the latter. There is also a lot of commented out draft code which should be removed later, right now I am keeping to to see if it has value as I progress.

### CODE REVIEW

**Comment by siman** [comment]
Methods names `update_category` and `edit_post_text` look inconsistent.
Both of them updated an entity, but the first one uses the word `update`, while the second `edit`

**Comment by bedeho** [comment]
Its not ready for review yet, those WIP commits will be removed.

File: src/lib.rs [hunk 1]
```diff
@@ -241,14 +259,17 @@ pub trait ForumUserRegistry<AccountId> {
 }
 
 /// Convenient composite time stamp 
-struct BlockchainTimestamp<BlockNumber, Moment> {
+//#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[derive(Encode, Decode, Default, Clone, PartialEq, Eq)]
+pub struct BlockchainTimestamp<BlockNumber, Moment> {
```
Comment (ghost) [inline src/lib.rs:344]: We have something a bit like that in the storage module; it'd be great to extract this into something more re-usable.

There are a few moments where we want to record the block number/time.

File: src/lib.rs [hunk 2]
```diff
@@ -208,9 +208,108 @@
...
+struct InputValidationLengthConstraint {
+
+    /// Minimum length
+    min : usize,
+
+    /// Maximum length
+    /// While having max would have been more direct, this
+    /// way makes max < min unrepresentable semantically, 
+    /// which is safer.
+    max_as_min_diff: usize,
```
Comment (siman) [inline src/lib.rs:237]: Cannot say that I like this idea of diff. Looks like an overenginering to me. 
1. You afraid that somebody silently will change max so it will be smaller then min. But an attacker can otherwise change plus sign to minus in this code: `self.min + self.max_as_min_diff`
2. We never used such approach in any other module.
3. I haven't seen this approach in other software.
4. UI should replicate the same function for `self.min + self.max_as_min_diff`.
5. You can check that min < max in validation function if you want.

File: src/lib.rs [hunk 3]
```diff
@@ -208,9 +208,108 @@
...
+        self.min + self.max_as_min_diff
+    }
+
+    /// Just to give method interface to read min, like max
+    fn min(&self) -> usize {
+        self.min
+    }
+
+    fn validate(&self, length: usize) -> LengthValidationResult {
+
```
Comment (siman) [inline src/lib.rs:253]: You can verify that min < max in this method.

File: src/lib.rs [hunk 4]
```diff
@@ -208,9 +208,108 @@
...
+/// Constants
+const CATEGORY_TITLE: InputValidationLengthConstraint = InputValidationLengthConstraint{
+    min: 3,
+    max_as_min_diff: 30
+};
+
+const CATEGORY_DESCRIPTION: InputValidationLengthConstraint = InputValidationLengthConstraint{
+    min: 10,
+    max_as_min_diff: 100
```
Comment (siman) [inline src/lib.rs]: Even tweets in the past had 140 chars ;)

File: src/lib.rs [hunk 5]
```diff
@@ -208,9 +208,108 @@
...
+};
+
+/// The greatest valid depth of a category.
+/// The depth of a root category is 0.
+const MAX_VALID_CATEGORY_DEPTH: u32 = 3;
```
Comment (siman) [inline src/lib.rs]: This name sounds more user-friendly to me, IMHO :)
```suggestion
const MAX_NUMBER_OF_SUBCATEGORIES: u32 = 3;
```

File: src/lib.rs [hunk 6]
```diff
@@ -208,9 +208,108 @@
...
+
+/// Error messages for dispatchables
+/// Later perhaps make error message functions, to add parametization
+
+const DISPATCHABLE_ERROR_FORUM_SUDO_NOT_SET: &str = "Forum sudo not set.";
```
Comment (siman) [inline src/lib.rs]: `DISPATCHABLE_` prefix looks redundant. `ERROR_` prefix is more than eough.
```suggestion
const ERROR_FORUM_SUDO_NOT_SET: &str = "Forum sudo not set.";
```

File: src/lib.rs [hunk 7]
```diff
@@ -208,9 +208,108 @@
...
+const DISPATCHABLE_ERROR_ORIGIN_NOT_FORUM_SUDO: &str = "Origin not forum sudo.";
+
+const DISPATCHABLE_ERROR_CATEGORY_TITLE_TOO_SHORT: &str = "Category title too short.";
+const DISPATCHABLE_ERROR_CATEGORY_TITLE_TOO_LONG: &str = "Category title too long.";
+
+const DISPATCHABLE_ERROR_CATEGORY_DESCRIPTION_TOO_SHORT: &str = "Category description too long.";
+const DISPATCHABLE_ERROR_CATEGORY_DESCRIPTION_TOO_LONG: &str = "Category description too long.";
+
+const DISPATCHABLE_ERROR_PARENT_CATEGORY_DOES_NOT_EXIST: &str = "Parent category does not exist.";
+const DISPATCHABLE_ERROR_PARENT_CATEGORY_DELETED: &str = "Parent directly category deleted.";
```
Comment (siman) [inline src/lib.rs]: Seems like there is a typo in this text: `Parent directly category deleted`

File: src/lib.rs [hunk 8]
```diff
@@ -208,9 +208,108 @@
...
+const DISPATCHABLE_ERROR_PARENT_CATEGORY_ARCHIVED: &str = "Parent directly category archived.";
```
Comment (siman) [inline src/lib.rs]: Seems like there is a typo in this text: `Parent directly category archived `

File: src/lib.rs (hunk 1 above)
Comment (siman) [reply src/lib.rs:344]: 100% agreed.

File: src/lib.rs [hunk 9]
```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
     forum_sudo_creator: AccountId
```
Comment (siman) [inline src/lib.rs]: It could be just `owner_id` or `creator_id`. I mean that in future different user roles could be able to create categories. But calling this `forum_sudo_creator` restricts it to the forum sudo.

File: src/lib.rs [hunk 10]
```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
...
-        // Just a normal `enum`, here's a dummy event to ensure it compiles.
-        /// Dummy event, just here so there's a generic type that's used.
-        Dummy(B),
+        /// A category was introduced
+        CategoryCreated(CategoryId),
+
+        /// A category with given id was updated. 
+        /// The second argument reflects the new archival status of the category, if changed.
+        /// The third argument reflects the new deletion status of the category, if changed.
+        CategoryUpdated(CategoryId, Option<bool>, Option<bool>),
```
Comment (siman) [inline src/lib.rs:577]: Did you consider to create more granular events like?:
- `CategoryArchived`
- `CategoryUnarchived`
- `CategoryDeleted`
- `CategoryUndeleted`

Having a few sequential parameters of the same type (Option<bool>) could be confusing when consuming the event because event parameters are not named.

File: src/lib.rs [hunk 11]
```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
...
+                Self::build_category_path(parent_category_id, &mut category_path);
+
+                // If path is empty, it just means the parent did not exist,
+                // which means teh user provided an invalid parent category id.
+                ensure!(category_path.len() > 0,
+                    DISPATCHABLE_ERROR_PARENT_CATEGORY_DOES_NOT_EXIST
+                );
+
+                // Is parent category directly or indirectly deleted category
+                ensure!(!category_path.iter().any(|c:&Category<T::BlockNumber, T::Moment, T::AccountId>| c.deleted),
```
Comment (siman) [inline src/lib.rs]: It could be optimized: you can find any category that is deleted or archived in one iteration and exit if found.
In your case you do two completely similar iterations - that is an expensive operation in blockchains :)

File: src/lib.rs [hunk 12]
```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
...
+
+            // Validate description
+            ensure_category_description_is_valid(&description)?;
 
-        // The signature could also look like: `fn on_finalize()`
-        fn on_finalize(_n: T::BlockNumber) {
-            // Anything that needs to be done at the end of the block.
-            // We just kill our dummy storage item.
-            <Dummy<T>>::kill();
+            if let Some(parent_category_id) = parent {
```
Comment (siman) [inline src/lib.rs:645]: Consider extracting the whole `if` block to a separate function called for example `verify_parent_category(...)`.
It will improve/simplify readability of `create_category` function.

File: src/lib.rs [hunk 13]
```diff
@@ -371,246 +493,346 @@ struct Category<BlockNumber, Moment, AccountId> {
...
+            <CategoryById<T>>::get(category_id)
+        } else {
+            Err("Category does not exist.")
+        }
+
+        Err("dddd")
+    }
+*/
+
+    fn build_category_path(category_id:CategoryId, path: &mut Vec<Category<T::BlockNumber, T::Moment, T::AccountId>>) {
```
Comment (siman) [inline src/lib.rs]: Consider checking a number of recursive calls. It should be aligned with max category depth.

**Review by siman — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 1 above)
Comment (bedeho) [reply src/lib.rs:344]: Yes, I agree, but this is a standalone repo and crate, so I cannot reuse before we actually create self-contained package with all of the common substrate things we want to share across all of our modules. This may be of value to others also.

File: src/lib.rs (hunk 2 above)
Comment (bedeho) [reply src/lib.rs:237]: I don't know if I will keep this `InputValidationLengthConstraint`, but I should respond to some of these points. Most importantly, I am not sure if you are just objecting to the idea of using the `max_as_min_diff` approach or more broadly having even this whole struct, even if `max` was explicitly represented. 

I will assume it is the first case.

1. I am not sure what you mean by "an attacker", which apparently can alter the source. Anyone who can alter the source can do anything, so it's a moot point. The underlying point is to use the type system to make bad states unrepresentable, which is the entire point of type safety, and that is what this approach does. It makes it so that a developer cannot accidentally allow an invalid max/min combination, which could happen in many ways, depending on how these constraints are being set or changed (e.g. manually in code and fixed forever, or changed with calls from governance).

2. Ok.

3. This is because you are misunderstanding what the goal here is because this is done all the time. The simple choice of using an unsigned variable rather than a signed variable here is just the same basic point.

4. It will need to be able to read the values and interpret them correctly, yes. It may choose to explicitly store max/min in its own representation, but the whole question is what is actually a safer way to represent it. Whatever the answer is, it should probably be the same in both runtime and UI code, they should both be as safe/practical possible.

5. This is a runtime check, which is strictly less safe and more complicated to handle than having the type system statically make it impossible. You can still have a bad state in your system that you somehow must recover from, and you will  have to have some sort of error handling in the code calling `validate`, otherwise, your system will still fail.

In conclusion, I am not sure if I keep this struct - but I think it should just stay for now, I feel like `LengthValidationResult` is the part that may be overengineered, but even if I drop both, I still think it's better to use the basic idea in the representation of the constraint itself.

File: src/lib.rs (hunk 3 above)
Comment (bedeho) [reply src/lib.rs:253]: Yep.

File: src/lib.rs (hunk 4 above)
Comment (bedeho) [reply src/lib.rs]: Changing to that then :D

File: src/lib.rs (hunk 5 above)
Comment (bedeho) [reply src/lib.rs]: I think `MAX_NUMBER_OF_SUBCATEGORIES` is ambiguous, it can be interpreted as 

"In one category, you can only have max 3 categories inside it". 

This is why I use the term "depth", because it has a clear meaning in the context of a tree, it refers to multiple levels.

Changing to `MAX_CATEGORY_DEPTH`, I think its quite clear. Also comment explains

```
/// The greatest valid depth of a category.
/// The depth of a root category is 0.
```

Anyway, not sure what you mean by "user", but developers get it, end users don't read these names, they will have some nice explanation in Pioneer or similar :)

File: src/lib.rs (hunk 6 above)
Comment (bedeho) [reply src/lib.rs]: I was thinking if we have other error messages that are not in dispatchable, then it would be nice to differentiate, but I think its unlikely, removing.

File: src/lib.rs (hunk 7 above)
Comment (bedeho) [reply src/lib.rs]: Fixed.

File: src/lib.rs (hunk 8 above)
Comment (bedeho) [reply src/lib.rs]: Fixed.

File: src/lib.rs (hunk 9 above)
Comment (bedeho) [reply src/lib.rs]: This was fixed in response to a review of the PR #3 

https://github.com/Joystream/substrate-forum-module/pull/3/commits/0795f21083b0ba8df768faa09ebd321ed062b938

File: src/lib.rs (hunk 10 above)
Comment (bedeho) [reply src/lib.rs:577]: I think that is better solved by just having proper documentation of events, which we currently don't have. One could make a similar argument to adding four different dispatchables also, but I don't think its worth going back on now.

File: src/lib.rs (hunk 11 above)
Comment (bedeho) [reply src/lib.rs]: Yes this could be optimized, but recall that this will typically be very short iteration, its over category depth, so max 1-4 steps. So I opted for simplicity. 

But anyway, I will change this in #4 , so that its a bit more reusable, so that you don't need to do this inline so many different dispatchbles, and I will probably do it like you suggest then.

File: src/lib.rs (hunk 13 above)
Comment (bedeho) [reply src/lib.rs]: I dont understand what you mean by checking? the termination condition makes sure it ends when it should?

Also recall that, if we make the max depth updatable, e.g. by moderator or governance, then this actually will not even be true, as old category tree structure may not admit to new max value, when max is reduced later.

Comment (siman) [reply src/lib.rs]: Makes sense!

File: src/lib.rs (hunk 2 above)
Comment (siman) [reply src/lib.rs:237]: Here I am concerned about max_as_min_diff. The whole valuation structure and methods look fine.

File: src/lib.rs (hunk 12 above)
Comment (bedeho) [reply src/lib.rs:645]: 1. Its hard to to factor out any function for this scenario which is genuinely reusable, am I missing something? 

2. Or are you saying to move it out even if it's not reusable? 

3. I have given it a try, factored  out `ensure_can_add_subcategory_to_parent_at_path_tip`, also factored out `ensure_can_mutate_in_category` because it will need to be used for many other dispatchables.

File: src/lib.rs (hunk 2 above)
Comment (bedeho) [reply src/lib.rs:237]: Ok, good to know.

**Review by siman — APPROVED** [review]

**Comment by siman** [comment]
Should I merge this PR?
